Initial osteocar reference comparison (#1426)
AlexsLemonade/OpenScPCA-analysis: analyses/cell-type-osteosarcoma/exploratory-notebooks/02-compare-references.Rmd

---
title: "Compare OsteoCar references"
author: "Stephanie J. Spielman"
date: "`r Sys.Date()`"
output:
  html_notebook:
    toc: true
    toc_depth: 3
    code_folding: hide
---

> There are several different references in `OsteoCar`, and while we can generally map them to our sample propertoes, we have some sample which don't neatly line up.

Reviewer: ```suggestion
There are several different references in `OsteoCar`, and while we can generally map them to our sample properties, we have some samples which don't neatly line up.
```

Reviewer: Just bumping this spelling comment.

Author: Disappointed you don't like propertoes, but fine. 🩴 🍹

The goal of this notebook is to lightly compare some of the references using `SingleR`:
* patient primary vs metastatic (lung only) references
* xenograft primary vs metastatic references
As these references are quite large, they have to be aggregated before use.

We compare the references to one another to see how similar the cell type distributions and expression signatures are.

## Setup

```{r}
suppressPackageStartupMessages({
  library(SingleR)
  library(SingleCellExperiment)
})
bp_param <- BiocParallel::MulticoreParam(6)
set.seed(2026)
```

```{r}
repository_base <- rprojroot::find_root(rprojroot::is_git_root)

module_dir <- file.path(repository_base, "analyses", "cell-type-osteosarcoma")
ref_dir <- file.path(module_dir, "references")
```

## Analysis

We'll compare `Ann_Level2` between primary and metastatic references by first aggregating them and then running `SingleR::matchReferences()` and plotting the resulting heatmap.
This will tell us where cell types overlap/differ between references.

Note that we read in references when we use them and remove them promptly after aggregation, to save memory.

### Primary references

```{r}

> patient_prim <- readr::read_rds(file.path(ref_dir, "patient_prim_sce.rds"))

Reviewer: Before you dive into creating the trained references, it would be helpful to just see a summary of what cell types are present in each of the references. I want to know which cell types are present in both and which cell types are unique based on label alone. I also want to know which cell types are tumor cells vs not tumor cell types. It looks like we might have some different classes of tumor cells here?

patient_prim_aggr <- SingleR::trainSingleR(
  ref = patient_prim,
  labels = patient_prim$Ann_Level2,
  de.method = "wilcox",
  aggr.ref = TRUE,
  BPPARAM = bp_param
)
rm(patient_prim)

patient_mets <- readr::read_rds(file.path(ref_dir, "patient_mets_sce.rds"))
patient_mets_aggr <- SingleR::trainSingleR(
  ref = patient_mets,
  labels = patient_mets$Ann_Level2,
  de.method = "wilcox",
  aggr.ref = TRUE,
  BPPARAM = bp_param
)
rm(patient_mets)

gc()
```

In the heatmap below, the primary reference is columns and the metastasis reference is rows.

```{r}
matched <- matchReferences(
  patient_mets_aggr$ref, 
  patient_prim_aggr$ref, 
  patient_mets_aggr$labels$full, 
  patient_prim_aggr$labels$full 
)

> pheatmap::pheatmap(

Reviewer: Can we subset this to only show shared labels in the first place? I want to know if they have the same label in both references, are they similar. If they are different labels then they are probably distinct so we they just add noise to the heatmap and make it harder to interpret.

Reviewer: I actually don't think you need this heatmap at all. It's fine to keep, but if the labels don't match up then the value is going to be 0 anyways so it doesn't totally make sense to me to show both?

### Xenograft references

Note that there are many fewer cell types here, because the reference considers only human-derived tumor cells.
Also note, `Ann_Level2` and `Ann_Level3` are the same for the xenograft references, so this is the most fine-grained we can achieve.

```{r}
xeno_prim <- readr::read_rds(file.path(ref_dir, "xeno_prim_sce.rds"))
xeno_prim_aggr <- SingleR::trainSingleR(
  ref = xeno_prim,
  labels = xeno_prim$Ann_Level2,
  de.method = "wilcox",
  aggr.ref = TRUE,
  BPPARAM = bp_param
)
rm(xeno_prim)

xeno_mets <- readr::read_rds(file.path(ref_dir, "xeno_mets_sce.rds"))
xeno_mets_aggr <- SingleR::trainSingleR(
  ref = xeno_mets,
  labels = xeno_mets$Ann_Level2,
  de.method = "wilcox",
  aggr.ref = TRUE,
  BPPARAM = bp_param
)
rm(xeno_mets)

gc()
```

In the heatmap below, the primary reference is columns and the metastasis reference is rows.

```{r}
matched <- matchReferences(
  xeno_mets_aggr$ref, 
  xeno_prim_aggr$ref, 
  xeno_mets_aggr$labels$full, 
  xeno_prim_aggr$labels$full 
)
pheatmap::pheatmap(
  matched, 
  col=viridis::magma(500), 
  cluster_rows = FALSE, 
  cluster_cols = FALSE
)
```

> There's a 1:1 match of proliferative here only, but the noise around other cell types suggests that there are distinct expression profiles between shared cell types across these references.

Reviewer: Sure there's only a 1:1 match for proliferative, but I see some fairly strong signals for 4 out of the 6 tumor classes. It would be helpful to see a table of how many of each cell type is present in the original dataset alongside this just for added context. As in, how much of the population is assigned to this "Interactive" population vs the proliferative population?

What is not clear is how much of that is signal vs noise. 
The `Interactive` profile also appears to be completely distinct between these.

## Conclusions

> Given the differences between primary/metastasis references for both patients and xenografts, it seems worth doing some initial exploration of annotation performance with `SingleR` next.

Reviewer: I think I would still run these together as a single reference with distinct labels. So you would attach `primary-` or `met-` to each of the labels prior to running. Then you can see which labels are assigned and if you get predominantly primary or met or an even mix of both within a single sample. You could then compare that to running as a single reference with harmonized labels where you don't distinguish between primary and met when building the reference model.

---
title: "Compare OsteoCar references"
author: "Stephanie J. Spielman"
date: "`r Sys.Date()`"
output:
  html_notebook:
    toc: true
    toc_depth: 3
    code_folding: hide
---

There are several different references in `OsteoCar`, and while we can generally map them to our sample propertoes, we have some sample which don't neatly line up.

> For example, `OsteoCar` has patient primary and metastatic (lung only) references, but we have a sample from metastatic tissue in bone which differs from the metastatic reference. 

Reviewer: A question for the future, is it really just one sample or is it multiple samples?

Author: Definitely a typo, it's multiple......

```
> sample_df |> dplyr::filter(primary_or_metastasis=="Metastasis") |> dplyr::count(tissue_location)
# A tibble: 10 × 2
   tissue_location           n
   <chr>                 <int>
 1 Back                      2
 2 Femur                     2
 3 Kidney                    1
 4 Left Lung                 1
 5 Left upper lobe lung      1
 6 Lung                     29
 7 Paraspinal                1
 8 Right Lung                1
 9 Right upper lobe lung     2
10 Thigh                     1
```

The goal of this notebook is to lightly compare some of the references using `SingleR`:
* patient primary vs metastatic (lung only) references
* xenograft primary vs metastatic references

We compare the references to one another to see how similar the cell type distributions and expression signatures are.

> As these references are quite large, they have to be aggregated before use, but we'll also explore their raw data too.

Reviewer: Can you add a note here about what's in `Ann_Level1` vs `Ann_Level2`? That might help with some of the confusion I was experiencing (but have since resolved).

## Setup

```{r}
suppressPackageStartupMessages({
  library(SingleR)
  library(SingleCellExperiment)
  library(ggplot2)
  library(patchwork)
})
bp_param <- BiocParallel::MulticoreParam(6)
set.seed(2026)
```

```{r}
repository_base <- rprojroot::find_root(rprojroot::is_git_root)

module_dir <- file.path(repository_base, "analyses", "cell-type-osteosarcoma")
ref_dir <- file.path(module_dir, "references")
```

## Functions

```{r}

> make_barplot <- function(prim_coldata, mets_coldata, facet = TRUE) {

Reviewer: ```suggestion
make_barplot <- function(prim_coldata, mets_coldata) {
```

    
  prim_table <- prim_coldata |>
    as.data.frame() |>
    dplyr::count(Ann_Level1, Ann_Level2) |>
    dplyr::mutate(ref = "primary")
  
  full_table <- mets_coldata |>
    as.data.frame() |>
    dplyr::count(Ann_Level1, Ann_Level2) |>
    dplyr::mutate(ref = "metastasis") |>
    dplyr::bind_rows(prim_table) |>
    dplyr::group_by(ref) |>
    dplyr::mutate(prop = n / sum(n)) |>
    dplyr::ungroup()
  
  p_counts <- ggplot(full_table) + 
    aes(x = Ann_Level2, y = n, fill = ref) + 
    geom_col(position = position_dodge()) +
    scale_fill_brewer(palette = "Set1") +
    ggtitle("Counts") +
    theme_bw() + 
    theme(axis.text.x = element_text(angle = 30, hjust = 1))
  
  p_props <- ggplot(full_table) + 
    aes(x = Ann_Level2, y = prop, fill = ref) + 
    geom_col(position = position_dodge()) +
    scale_fill_brewer(palette = "Set1") +
    ggtitle("Proportions") +
    theme_bw() + 
    theme(axis.text.x = element_text(angle = 30, hjust = 1))
  
  if (facet) {
    p_counts <- p_counts + facet_wrap(vars(Ann_Level1), scales = "free", nrow=3)
    p_props <- p_props + facet_wrap(vars(Ann_Level1), scales = "free", nrow=3)

>   }

Reviewer: ```suggestion
  p_counts <- p_counts + facet_wrap(vars(Ann_Level1), scales = "free", nrow=3)
  p_props <- p_props + facet_wrap(vars(Ann_Level1), scales = "free", nrow=3)
 
```

It looks like you don't actually use the facet option, so you can probably remove it.

## Patient reference analysis

### Read references

We'll read in and aggregate, extract `colData` so we can ditch these objects as soon as possible.

```{r}
patient_prim <- readr::read_rds(file.path(ref_dir, "patient_prim_sce.rds"))
patient_prim_aggr <- SingleR::trainSingleR(
  ref = patient_prim,
  labels = patient_prim$Ann_Level2,
  de.method = "wilcox",
  aggr.ref = TRUE,
  BPPARAM = bp_param
)
prim_coldata <- colData(patient_prim)
rm(patient_prim)

patient_mets <- readr::read_rds(file.path(ref_dir, "patient_mets_sce.rds"))
patient_mets_aggr <- SingleR::trainSingleR(
  ref = patient_mets,
  labels = patient_mets$Ann_Level2,
  de.method = "wilcox",
  aggr.ref = TRUE,
  BPPARAM = bp_param
)
mets_coldata <- colData(patient_mets)
rm(patient_mets)

gc()
```

### Reference cell types

First, where do cell type labels overlap or not?

The primary reference uniquely has these cell types, both of which are mesenchymal:

```{r}
setdiff(
  unique(prim_coldata$Ann_Level2), unique(mets_coldata$Ann_Level2)
)
```

The metastasis reference uniquely has these cell types, across several categories.

```{r}
setdiff(
  unique(mets_coldata$Ann_Level2), unique(prim_coldata$Ann_Level2)
)
```

> The majority of cell types, including all classes of tumor cells, appear to be shared, but this does not guarantee they have similar expression.

Reviewer: It's been a while since I've read this paper, but can you remind me exactly which of these cell types they are calling "tumor" cell types vs. normal cell types? I'm actually surprised that the osteoblasts and chondrocytes population aren't overlapping with tumor, so I want to double check thats the case? Or is tumor vs normal represented by a different column in the data?

```{r}
shared <- intersect(
  unique(mets_coldata$Ann_Level2), unique(prim_coldata$Ann_Level2)
)
shared
```

We'll visualize this as well with barplots, faceted by cell type category, of cell type counts and proportions for these references for complementary views on their distributions.

```{r}
#| fig.height: 8
#| fig.width: 10

> make_barplot(prim_coldata, mets_coldata) |>

Reviewer: Ah I see now which categories are which. Thank you for adding this since it's helpful to see them grouped by broad cell type.

  wrap_plots(ncol=2, guides = "collect")
```

We see:

* Relatively more endothelial in mets, and no epithelial in primary
* Same immune lymphoid cell types but primary has relatively more T cells in the reference
* Similar immune myeloid distributions but primary has relatively more TAMs
* Most differences fall out in the mesenchymal cell types
* Same sets of tumor cells, but some different enrichments between references

### Heatmap 

We'll make a heatmap comparing expression across references.

We'll compare `Ann_Level2` between primary and metastatic references by first aggregating them and then running `SingleR::matchReferences()` and plotting the resulting heatmap.
This will tell us where cell type expression patterns overlap/differ between references.

```{r}
matched <- matchReferences(
  patient_mets_aggr$ref, 
  patient_prim_aggr$ref, 
  patient_mets_aggr$labels$full, 
  patient_prim_aggr$labels$full 
)
```

First, we'll show all cell types:

```{r}
pheatmap::pheatmap(
  matched, 
  col=viridis::magma(500), 
  cluster_rows = FALSE, 
  cluster_cols = FALSE
)
```

Now, only the cell types shared between the references:

```{r}

> shared_matched <- matched[shared, shared]

Reviewer: This is much more helpful because you can see the cell types that don't have any match that should have a match, which look to be mostly tumor cells. This tells me there probably is a difference in the marker gene expression for tumor cell subpopulations between primary and mets that we should be sure to consider.
